data-access/Helpers/Exceptions.cs

```diff
@@ -1,38 +1,151 @@
+    // response codes.  This enables a RESTful experience regardless of whether a request is made via HTTP or by
+    // websocket.  In either case the HTTP router or the Hub will catch a child of the HttpException class
+    // and return that information to the user.
+    // Errors should be thrown properly at the lowest levels of code, this will reduce boiler-plate in higher
+    // level functions like the Services and ensure more consistent responses.
+    public class HttpException : Exception
     {
-        protected RepositoryException(string msg) : base(msg) { }
-    }
+        private readonly int httpStatusCode;
 
-    public class NoPermissionException: RepositoryException
-    {
-        public NoPermissionException(uint? userId, string operation, string entity, uint? entityId) : 
-            base($"User ${userId?.ToString() ?? "anonymous"} can't perform ${operation} on ${entity} ${entityId?.ToString()}")
-        { }
-    }
-    
-    public class ImproperRequestException: RepositoryException
-    {
-        public ImproperRequestException(string operation, string requestError) : 
-            base($"Request ${operation} failed; ${requestError}.")
-        { }
+        public HttpException(HttpStatusCode httpStatusCode, HttpExceptionMessage msg) : base(JsonConvert.SerializeObject(msg))
+        {
+            this.httpStatusCode = (int)httpStatusCode;
+        }
+        
+        public HttpException(HttpStatusCode httpStatusCode, uint internalErrorCode, string msg = null) 
+            : base(JsonConvert.SerializeObject(new HttpExceptionMessage(internalErrorCode, msg)))
+        {
+            this.httpStatusCode = (int)httpStatusCode;
+        }
+
+        public int StatusCode { get { return this.httpStatusCode; } }
     }
-    
-    public class DbFailedWrite: RepositoryException
+
+    public class HttpExceptionMessage
     {
-        public DbFailedWrite() : 
-            base($"Failed writing to DB.")
-        { }
+        public uint internalErrorCode { get; set; }
+        public string message { get; set; }
+        
+        public HttpExceptionMessage(uint internalErrorCode, string message = null)
+        {
+            this.internalErrorCode = internalErrorCode;
+            this.message = message;
+        }
+
     }
     
-    public class DbDetailedFailedWrite: RepositoryException
+    // This is a collection of ready-made errors that can be thrown. They include a standard HTTP status error code
+    // and an internal project error code with accompanying message in English.
+    public static class StandardErrors
     {
-        public DbDetailedFailedWrite(string msg) : 
-            base($"Failed writing to DB. {msg}")
-        { }
+
+        #region Permissions errors
+        public static HttpException NoPermissions(uint userId, uint editionId)
+        {
+            return new HttpException(HttpStatusCode.Forbidden, 620, 
```

Comment:
We probably need a spreadsheet with all the errors and error codes. It'll be easier when we create an i18ned frontend again.

Reply:
Good idea!  Let me start a Googledocs spreadsheet.  Do you have a preference for spacing things out and maybe creating code domains? I left some small holes, but they are not too big.

Reply:
I prefer having codes as strings, anyway, so you can understand what the error is from the code itself.

Reply:
Ah, so are you suggesting that we remove the internal code numbers altogether and just use a descriptive English string as the code?

Reply:
Yes. NO_PERMISSIONS is a very good code for a NoPermissions exception.